cpp/src/reductions/simple_segmented.cuh

```diff
@@ -103,6 +112,112 @@ std::unique_ptr<column> simple_segmented_reduction(column_view const& col,
+ * the reduced value is valid if any element in the segment is valid.
+ * @param stream Used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return Output column in device memory
+ */
+
+template <typename InputType,
+          typename Op,
+          CUDF_ENABLE_IF(std::is_same_v<Op, cudf::reduction::op::min> ||
+                         std::is_same_v<Op, cudf::reduction::op::max>)>
+std::unique_ptr<column> string_segmented_reduction(column_view const& col,
+                                                   device_span<size_type const> offsets,
+                                                   null_policy null_handling,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::mr::device_memory_resource* mr)
+{
+  // Pass to simple_segmented_reduction, get indices to gather, perform gather here.
+  auto device_col = cudf::column_device_view::create(col, stream);
+
+  auto it                 = thrust::make_counting_iterator(0);
+  auto const num_segments = static_cast<size_type>(offsets.size()) - 1;
+
+  bool constexpr is_argmin = std::is_same_v<Op, cudf::reduction::op::min>;
+  auto string_comparator =
+    cudf::detail::element_argminmax_fn<InputType>{*device_col, col.has_nulls(), is_argmin};
+  auto constexpr identity =
+    is_argmin ? cudf::detail::ARGMIN_SENTINEL : cudf::detail::ARGMAX_SENTINEL;
+
+  auto gather_map =
+    cudf::reduction::detail::segmented_reduce(it,
+                                              offsets.begin(),
+                                              num_segments,
+                                              string_comparator,
+                                              identity,
+                                              stream,
+                                              rmm::mr::get_current_device_resource());
+  auto result = std::move(cudf::detail::gather(table_view{{col}},
+                                               *gather_map,
+                                               cudf::out_of_bounds_policy::NULLIFY,
+                                               cudf::detail::negative_index_policy::NOT_ALLOWED,
+                                               stream,
+                                               mr)
+                            ->release()[0]);
+  auto const [segmented_null_mask, segmented_null_count] =
+    cudf::detail::segmented_null_mask_reduction(col.null_mask(),
+                                                offsets.begin(),
+                                                offsets.end() - 1,
+                                                offsets.begin() + 1,
+                                                null_handling,
+                                                stream,
+                                                mr);
```

Comment:
A super minor nit: we can determine if `segmented_null_mask` is a temporary memory based on `result->null_count()` and choose to use default resource v.s. supplied resource here.

Reply:
Feel free to ignore or punt to next one @bdice.

Reply:
We are already using `mr` as efficiently (minimally) as possible. If `result->null_count() == 0`, we return the `mr`-allocated bitmask. Else, we do an `inplace_bitmask_and` that alters the `mr`-allocated bitmask before returning it.

The use of `mr` in `inplace_bitmask_and`, which calls `inplace_bitmask_binop`, is a little misleading. The allocator there is always a temporary allocator and is not part of the return value. I think it may be possible to refactor the `inplace` functions to remove `mr` as an argument, and require the use of the default memory resource to avoid an implication that it is used for the returned data (it is not, since it acts in-place on bitmasks passed in that were from some other allocator).

Reply:
When the gather result contains null, its `null_mask` is allocated with `mr`. The `inplace_bitmask_and` uses this as the target null mask, in this situation, the `segmented_null_mask` doesn't need to use the supplied `mr`.

Reply:
Okay, that might be the case. The `segmented_null_mask_reduction` could use `result->null_count() == 0 ? mr : rmm::mr::get_current_device_resource()`. I'm not sure if I would recommend using a conditional `mr` for the sake of clarity, but it seems that it could be done.